tendant/graphql-clj
# Post validation transformations

Various bugfixes and optimizations:
- Add test for execution on field arguments with variable bindings
- Build arguments, expand fragments, and inline types prior to execution phase
- Refactor executor to be simpler (delete the majority of the code)
- Warnings for library users accidentally doing parsing / validation in the inner loop
- Bugfix - break on undefined variables, don't traverse deeper
- Bugfix - add test for missing required variables, and support default values for variables
- Bugfix - guard against null values for non-null types in the execution phase
- Bugfix - support execution with inline fragments
- Bugfix - validation for missing root fields
- Bugfix - required arguments fulfilled by required variables
- Add preparation and memoization as an example in executor test
- Optionally add a `:resolver` key to state when validating a statement to inline resolver-fns prior to execution (and add example to executor test case)
- Streamline statement validation output - remove :state key in the absence of errors
- Add a spec for statement validation output and ensure test cases conform
- Add starwars test case from the `graphql-clj-starter project`
- Bugfix for introspection resolution - isDeprecated is a non-null field

Addresses:
* https://github.com/tendant/graphql-clj/issues/42
* https://github.com/tendant/graphql-clj/issues/36 - doesn't change the current approach, but removes much of the complexity from it
* https://github.com/tendant/graphql-clj/issues/35
* https://github.com/tendant/graphql-clj/issues/43 - partial fix - relaxes validation constraint for now until we can resolve object equality for these specs - better than blocking valid statements in the meantime.

Related: 
* https://github.com/tendant/graphql-clj-starter/pull/1
* https://github.com/xsc/graphql-clj-bench/pull/2
